### ISSUE TITLE
Change Flutter TODOs to not require a username, but to require an issue link.

### ISSUE THREAD

**Issue by gspencergoog** [flutter/flutter#37519]
The old syntax was:
```
// TODO(username): explanation, https://link/to/issue
```
But the link wasn't required.

The new syntax is:
```
// TODO: explanation, https://link/to/issue
```
Where the username is optional, and the link is required.

**Comment by gspencergoog** [comment]
cc @mklim

**Comment by mklim** [comment]
Also see dart-lang/linter#1670.

This is going to require multiple steps:

- [ ] Turn off the current `flutter_style_todos` lint, or make it more permissive.
- [ ] Update existing TODOs in the codebase to the new syntax. This will require filing issues for the ones currently missing links.
- [ ] Apply a lint for the new style.

2 and 3 will need to happen very close to each other or risk failing on new TODOs.

**Comment by jmagman** [comment]
Additional step: update the [style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#comments).

**Comment by chunhtai** [comment]
@gspencergoog Do we still want to make the user name optional?

**Comment by gspencergoog** [comment]
Yes, I think so. I've certainly wanted to add TODOs where I'm not the one with the knowledge of what's going on (and don't know who does), but I have a link to an issue.

**Comment by chunhtai** [comment]
In that case, those todos will be orphans and (likely) no one would ever pick it up.

One of the action item for my todo audit is to make sure every todo has an owner, and Owners will receive email notification periodically to review the current TODO in the repository. The goal is to make sure people are aware of the tech debts they added to the repository.

If you don't have knowledge about the context around the TODO, how do you know there is something need to do?

**Comment by gspencergoog** [comment]
The TODO username isn't supposed to be the "owner" of the task of removing that TODO: a TODO is not a commitment that the person referenced will fix the problem. They are supposed to be someone who has context about the reason for adding that TODO. They might also happen to be the owner of the task of fixing it, but not necessarily. Which is part of why issues should be required, so we can see who owns the issue (and we can reassign it without changing all the TODOs).

Sometimes I want to add a TODO where the context resides with an issue instead of an individual. For example, something where the entire team or a large portion of the team is supposed to have the context (e.g. null-safety migration), or something where another group will have the context (e.g. Dart team). I suppose I could add my name to one of those, but if I'm only one of many, it doesn't make a lot of sense to add me to it instead of (for example) you, other than I'm the one that added that particular instance.

I suppose requiring adding a name to it isn't the end of the world: I did add the TODO, afterall, but sometimes it would just mean if asked I would just redirect someone to the issue because I have only a small idea of the context.

**Comment by jmagman** [comment]
Also, names get out of date as people leave the team.

https://github.com/flutter/flutter/blob/e7b7ebc066c1b2a5aa5c19f8961307427e0142a6/packages/integration_test/example/integration_test/_extended_test_io.dart#L29
https://github.com/flutter/flutter/blob/734df6f5c6c0aaaebf8fbbca72592be3c5a17b91/packages/flutter/lib/src/widgets/platform_view.dart#L169

**Comment by mklim** [comment]
In addition to what @gspencergoog and @jmagman have said, linking to issues mean that context on the problem is available for anyone reading through the code, and they can look through the issue to see its status. Just linking to a username means that any reader would need to find the person and chat with them to understand anything unusual around the TODO, and how it's being prioritized.

**Comment by chunhtai** [comment]
@jmagman Another goal is to make sure obsolete TODO is reassigned. I see the argument, but that make the username not as useful if eventually people are just going to checkout the issue instead, and as @jmagman said, the username will be come out of date eventually.

@mklim My original plan is to enforce both username and issue link need to be present for the TODO, so people still have a link to the issue not just a username.

We enforce the username requirement as of now and the number of TODOs is already quite big, I am worrying that relaxing that requirement will make it even worse as people will have less burden when they decide to file a nameless TODO.

**Comment by gspencergoog** [comment]
I think a worse outcome than nameless TODOs would be that people add fewer TODOs so as to not incur a burden. That's why currently adding a TODO doesn't include the commitment to fix the TODO.

**Comment by gspencergoog** [comment]
The "nameless" TODOs are only nameless in the codebase: they would still be required to have an issue associated with them, and those issues all have names.

**Comment by chunhtai** [comment]
cc @gw280

**Comment by Hixie** [comment]
I don't think we should do this.

The point of having the name in the TODO isn't that you're responsible for it, or that you're the contact point, it's to make you think twice about adding the technical debt in the first place because you know that your name will be associated with it until it's fixed.

**Comment by Hixie** [comment]
(In particular, I don't think there's any value in "reassigning" TODOs -- they're not assigned in the first place.)

**Comment by chunhtai** [comment]
@Hixie

1. Make name optional
2. Male issue link required

I assume you disagree with (1), not both (1) and (2), right?

Right now the name is required, I plan to enforce the issue link first, and then we can decide whether to do (1)

**Comment by Hixie** [comment]
I'm ambivalent about (2), if people think that it's helpful to have issues for TODOs then I'm fine with it.

**Comment by Hixie** [comment]
I was asked for an update about my opinions on this on Discord, but for completeness let me add my reply here too:

I don't have a strong opinion about whether or not to do this change so long as:
1. we migrate all the TODOs when we change this over (promptly, not over multiple weeks or months)
2. we're consistent across our repos
3. we have tooling that supports our choice
4. the syntax isn't ugly (e.g. i'd prefer `TODO: text, https://url` over `TODO(https://url): text`)

Having the username in the TODO I think has significant psychological value in that it causes most people to think carefully about whether they want to add a TODO or whether they want to just fix the problem in the first place; it also makes it easier to know who to talk to when you trace a bug right to a TODO (which isn't uncommon for well placed TODOs).

Requiring an issue link _and_ a TODO has some additional cost (e.g. increases our issue count artificially, makes it annoying when writing code as you can't really file the issue until the PR lands but also you don't know if the TODO will stay in the PR until it lands, etc; makes it harder for us to easily tell when we can just close an issue, and means we have to keep them in sync -- this gets even worse when there's not a 1:1 mapping).

In general I count also support a world where we just don't add TODOs at all. I often will remove TODOs that have been in the codebase for a while and aren't providing much value. The ones that do provide value could often just be written as normal comments.

I suspect if we require a URL then people will just write TODOs without the word "TODO" when they want to leave one without the overhead of creating an issue.

**Comment by gspencergoog** [comment]
I agree with all four of Ian's "so long as" points above.

For my thought process: I was thinking that the reluctance that people might have to putting their name by it might be causing us to not be aware of some of the technical debt in our system because people just don't add a TODO because they feel intimidated by adding their name to it explicitly (Of course, they're also already associated with a comment line in the code via the blame list, it's just slightly harder to see).

I do see Ian's point about requiring an issue being a stumbling block along the same lines: you might not put in a TODO at all if it's annoying to create the issue (and it often is). It just seems like an issue has the right attributes that I was looking for: something where we can put more detail and discuss, that can be updated over time, and reassigned and given priorities.

Making adding a TODO psychologically easier to do seems like a good idea to me, as long as there's a chance we can understand more details about it someday to fix it and/or remove it. I mean, unless the goal is just to keep the number of TODOs down to make our tech debt numbers look good. :-)

We could just settle for recommending adding an issue, but not requiring it, I guess. But then there's precious little information in a TODO other than the description of the PR that added it (and that's dated, and probably won't be updated).

We could just not add TODOs at all and instead make the policy "If you were going to add a TODO, if you think it's important, just file an issue with the tech-debt label instead" and then we track it from then on as an issue. It makes it harder to know to remove it, though, if it's not linked to the pertinent code, although the location of a TODO already often ends up not having much to do with the location of its solution.

So, I guess my preference is still that we require links but not names, because that seems to me like it has the most value in the long term. If that's unpopular, then I think I'd just choose to get rid of TODOs and recommend that people add implementation comments and file issues instead (perhaps referenced from the comment if it seems helpful).

**Comment by matanlurey** [comment]
Sounds good. I'm going to chat with the analyzer time on how to best contribute a change now that the linter lives in `dart-lang/sdk`. My goal is to make this as painless as possible, so likely something like:

- [ ] Allow both styles (name and nameless with URL)
- [ ] Work with the larger team to delete or migrate to the newer style (there are ~300 usages in `flutter/flutter`)
- [ ] Disallow the older style

Hopefully the time between "allow" and "disallow" is relatively low (measured in ~weeks not ~quarters).

**Comment by matanlurey** [comment]
But this is our own lint, if we can't introduce changes to our own lint then what was the point?

I could imagine writing `flutter_style_todos_2` (probably better named as `prefer_todos_with_links`) that temporarily allows both styles (i.e. the same story as above, just with a different line name). Is that better?

**Comment by matanlurey** [comment]
That's sort of the industry standard for making these kinds of changes, but I won't decline someone else doing work for me.

**Comment by bwilkerson** [comment]
> But this is our own lint, if we can't introduce changes to our own lint then what was the point?

No, it isn't our own lint. Yes, it's a lint added to enforce the style of TODOs the Flutter team was using at the time, but it was published as a lint available to everyone in the Dart and Flutter community, and some of them are using it.

If you want the ability to have a set of lints used by the Flutter team that nobody else can enable for their code, then we'll need to talk about how to make that possible. At the moment isn't an option.

**Comment by matanlurey** [comment]
Gotcha, my mistake then

**Comment by natebosch** [comment]
Internal style recommends `// TODO: $LINK - $DETAILS`. If we are migrating to a new style I think aiming for consistency is the best default choice. One small difference though is that internal links tend to be short. The length of a full github URL could be a good reason to push the link to the end of the line.

The internal guide also prefers issue links to usernames. I do not think adding a username to a TODO is likely to be a meaningful incentive for good engineering practices. Issue links are more useful to a future reader than a `git blame` shortcut. A thoughtful TODO description is more useful than a username and vague description.

I think it would be OK either way to allow or disallow TODOs with no links. If we allow a TODO with no link a reviewer should hopefully check that the TODO has both a clear description of the anticipated change and when we should make that change. The style guide should focus more on whether a TODO is clearly actionable than whether it repeats `git blame` info.

**Comment by matanlurey** [comment]
It doesn't sound like we have consensus over the solution then.

I also agree with Nate that it's unlikely we could fit most URLs on one line, meaning the lint would have to permissive for overflows, i.e.:

```dart
// TODO: a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a.
// https://github.com/flutter/flutter/issues/12345.
```

At that point, not only does this lint become hard to write (how many lines, 2? what about 3?), but it sounds like we're likely to diverge and just create another lint style that nobody is really happy with.

My suggestion would be to just drop the lint we have now and have no lint, but I think others would strongly push back, and I don't have the bandwidth to continue to push, so I'll keep this issue open for the next person.
